src/query/storage/cache/redis_cache.go

```diff
@@ -262,18 +255,171 @@ func WindowGetOrFetch(
+	// Since otherwise getting the last bucket and storing it could avoid storing new data
+
+	// If we have more than one bucket, we want to get every bucket from the last from M3DB
+	if q.Start.Before(last_start) {
+		// Split into queries of the last bucket and everything else
+		exclude_last_query := &storage.FetchQuery{
+			TagMatchers: q.TagMatchers,
+			Start:       q.Start,
+			End:         last_start,
+			Interval:    q.Interval,
+		}
+		last_query := &storage.FetchQuery{
+			TagMatchers: q.TagMatchers,
+			Start:       last_start,
+			End:         q.End,
+			Interval:    q.Interval,
+		}
+		buckets := splitQueryToBuckets(exclude_last_query, BucketSize)
+		// Check if we have to make a request to M3DB, in which case we just ask for all
+		// This is so we don't have to convert if we need to ask M3DB for all of it anyways
+		count := cache.Check(buckets, BucketKeyPrefix)
+		cache.logger.Info("num_hit", zap.Int("check", count), zap.Int("length", len(buckets)))
+		// We need to get from M3DB when we are missing more than total allowed missing buckets
+		// Or when we have no buckets (in the case where we only want 1 bucket, and nothing is there)
+		if count < len(buckets)-AllowedMissingBuckets || count == 0 {
+			cache.logger.Info("cache miss", zap.String("key", KeyEncode(q, BucketKeyPrefix)))
+			res, err := st.FetchProm(ctx, q, fetchOptions)
+			cache.cacheMetrics.CacheMetricsMiss(res)
+			if err == nil {
+				cache.SetAsBuckets(&res, buckets)
+			}
+			return res, err
+		}
+
+		// Get the data for each bucket
+		results := cache.Get(buckets, BucketKeyPrefix)
+		cnt := 0
+		// Check again after getting just in case some of the keys disappeared so we don't make many M3DB requests
+		for _, r := range results {
+			if r == nil {
+				cnt++
+				// In case we end up missing more than the allowed amount of missing buckets
+				if cnt > AllowedMissingBuckets {
+					cache.logger.Info("cache miss", zap.String("key", KeyEncode(q, BucketKeyPrefix)))
+					res, err := st.FetchProm(ctx, q, fetchOptions)
+					cache.cacheMetrics.CacheMetricsMiss(res)
+					if err == nil {
+						cache.SetAsBuckets(&res, buckets)
+					}
+					return res, err
+				}
+			}
+		}
+		// We iterate again since in the above iteration, we just do an iteration to check if all the data is still there
+		// Allows us to avoid making more than the acceptable number of M3DB requests
+		for i, r := range results {
+			if r == nil {
+				res, err := st.FetchProm(ctx, buckets[i], fetchOptions)
+				if err != nil {
+					cache.logger.Error("M3DB Fetch Error of Bucket", zap.Error(err))
+					return storage.PromResult{}, nil
+				}
+				cache.cacheMetrics.CacheMetricsMiss(res)
+				results[i] = &res
+				cache.Set([]*storage.FetchQuery{buckets[i]}, []*storage.PromResult{results[i]}, BucketKeyPrefix)
+			}
+			// If it's the first bucket, we need to filter it to the actual start of query
+			if i == 0 {
+				filterResult(results[i], q.Start.Unix())
+			}
```

Comment:
not sure I follow this line, can you put more comments?

Reply:
added